fitnesstrax: init at 0.1.0

savannidgerinel opened the pull request:
###### Motivation for this change

This adds a fitness tracking application that I am developing.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

savannidgerinel committed:
Add a derivation for Fitnesstrax

This is a fitness tracking application that I am developing.

worldofpeace commented on pkgs/applications/misc/fitnesstrax/default.nix:
```diff
@@ -0,0 +1,44 @@
+  stdenv ? pkgs.stdenv,
+  rustPlatform ? pkgs.rustPlatform,
+  wrapGAppsHook ? pkgs.wrapGAppsHook,
+}:
```
```suggestion
{ pkgs
, fetchFromGitHub
,  stdenv
,  rustPlatform
, wrapGAppsHook
}:
```

callPackage is going to fill this in properly already

savannidgerinel replied:
Done, plus added the other formal dependencies.

worldofpeace commented on pkgs/applications/misc/fitnesstrax/default.nix:
```diff
@@ -0,0 +1,44 @@
+    pkgs.gtkd
+    pkgs.gtk3
+    wrapGAppsHook
+    ];
```
please reference these formally (without pkgs). All of these except `wrapGAppsHook` should be `buildInputs` also.

savannidgerinel replied:
Split. only needed wrapGAppsHook in nativeBuildInputs, and the rest in work in buildInputs.

worldofpeace commented on pkgs/applications/misc/fitnesstrax/default.nix:
```diff
@@ -0,0 +1,44 @@
+    mkdir -p $out/share/glib-2.0/schemas
+    cp -r $src/share/* $out/share/
+    glib-compile-schemas $out/share/glib-2.0/schemas
+    '';
```
```suggestion
  '';
```

worldofpeace commented on pkgs/applications/misc/fitnesstrax/default.nix:
```diff
@@ -0,0 +1,44 @@
+    description = "Privacy-first fitness tracking. Your health, your data, your machine.";
+    homepage = "https://github.com/luminescent-dreams/fitnesstrax";
+    license = licenses.bsd3;
+    maintainers = [ "savanni@luminescent-dreams.com" ];
```
can you add yourself to `maintainers-list.nix`?

worldofpeace commented on pkgs/applications/misc/fitnesstrax/default.nix:
```diff
@@ -0,0 +1,44 @@
+    '';
+
+  meta = with stdenv.lib; {
+    description = "Privacy-first fitness tracking. Your health, your data, your machine.";
```
```suggestion
    description = "Privacy-first fitness tracking. Your health, your data, your machine";
```

the docs on meta attributes in nixpkgs say no ending period

worldofpeace commented on pkgs/applications/misc/fitnesstrax/default.nix:
```diff
@@ -0,0 +1,44 @@
+  wrapGAppsHook ? pkgs.wrapGAppsHook,
+}:
+rustPlatform.buildRustPackage rec {
+  name = "fitnesstrax-${version}";
```
```suggestion
  pame = "fitnesstrax";
```

worldofpeace commented on pkgs/applications/misc/fitnesstrax/default.nix:
```diff
@@ -0,0 +1,44 @@
+  src = fetchFromGitHub {
+    owner = "luminescent-dreams";
+    repo = "fitnesstrax";
+    rev = "516a9e48978ac0f90bbbeade34d626211ccb78a3";
```
do you have a tag in your repo? then you could do `rev = version;`

savannidgerinel replied:
Sweet. I did not know that was possible.

savannidgerinel replied:
Tagged at fitnesstrax-0.1.0!

worldofpeace commented:
Why doesn't your makefile do what is done in `postInstall`. I could see if I could find a project that does exactly this as an example.

savannidgerinel commented:
> Why doesn't your makefile do what is done in `postInstall`? I could see if I could find a project that does exactly this as an example.

I haven't seen makefiles in Rust projects before, but I've also not seen much in the way of Rust-based executables. Perhaps this is just a place where the ecosystem doesn't have sufficient known standards?

If you have an example, I would definitely follow it. Otherwise, I thought I had seen in my experiments that the shared/ folder just doesn't get copied into the destination.

worldofpeace commented:
> Why doesn't your makefile do what is done in `postInstall`? I could see if I could find a project that does exactly this as an example.

GLib provides https://gitlab.gnome.org/GNOME/glib/-/blob/master/m4macros/gsettings.m4, but you 'd need to use autotools. In that case I would really just recommend using meson.

worldofpeace commented:
Oh, I thought you were using makefile somehow from looking at your project. Nevermind that then.

worldofpeace commented:
I think you might want to delete this file from your repo https://github.com/luminescent-dreams/fitnesstrax/blob/sol/share/glib-2.0/schemas/gschemas.compiled. It's compiled but not meant to be distributed in that sense.

savannidgerinel committed:
Changes from code review

savannidgerinel committed:
Set the correct version tag and the corresponding cargo sha

savannidgerinel commented:
> I think you might want to delete this file from your repo https://github.com/luminescent-dreams/fitnesstrax/blob/sol/share/glib-2.0/schemas/gschemas.compiled. It's compiled but not meant to be distributed in that sense.

I'll do that in the next release. I recall needing it to get the derivation right, but it may have been an intermediate step while I was flailing.

savannidgerinel committed:
Fix a typo in the maintainer list

savannidgerinel commented:
Having addressed the changes, @worldofpeace and I get different sha256sums on the cargo file. I'm going to hold off until a few days and try stabilizing things on nixos-unstable, as suggested in IRC.

worldofpeace commented on maintainers/maintainer-list.nix:
```diff
@@ -8297,6 +8297,12 @@
+    github = "savannidgerinel";
+    githubId = 8534888;
+    name = "Savanni D'Gerinel";
+  };
```
I believe this has to be alphabetical?

savannidgerinel replied:
Done. That wasn't entirely clear, but it looks like the end of the file has just gotten out of order.

worldofpeace commented on pkgs/applications/misc/fitnesstrax/default.nix:
```diff
@@ -0,0 +1,51 @@
+    gnome2.pango
+    cairo
+    glib
+    gdk_pixbuf
```
```suggestion
    gdk-pixbuf
```

that's an old alias, it was renamed

worldofpeace commented on pkgs/applications/misc/fitnesstrax/default.nix:
```diff
@@ -0,0 +1,51 @@
+    cairo
+    glib
+    gdk_pixbuf
+    gtkd
```
where do you use dlang bindings?

savannidgerinel replied:
Sorry, I don't understand the question. What do you mean?

worldofpeace replied:
gtdk is `D binding and OO wrapper for GTK` https://dlang.org/. Isn't this all in rust?

savannidgerinel replied:
Going to guess that was a leftover of some early flailing as I was trying to chase down all of the dependencies, and that I forgot about it.

worldofpeace commented on pkgs/applications/misc/fitnesstrax/default.nix:
```diff
@@ -0,0 +1,51 @@
+
+  buildInputs = [
+    atk
+    gnome2.pango
```
:( sucks that you have to use the old version of pango

savannidgerinel replied:
Yeah, doesn't appear that there's a new derivation. But since I don't use it directly, I don't so much notice.

worldofpeace commented on pkgs/applications/misc/fitnesstrax/default.nix:
```diff
@@ -0,0 +1,51 @@
+    description = "Privacy-first fitness tracking";
+    homepage = "https://github.com/luminescent-dreams/fitnesstrax";
+    license = licenses.bsd3;
+    maintainers = [ "savanni@luminescent-dreams.com" ];
```
```suggestion
    maintainers = [  maintainers.savanni ];
```

savannidgerinel committed:
Changes from code review

savannidgerinel committed:
Trying to fix the maintainers meta

worldofpeace commented on pkgs/top-level/all-packages.nix:
```diff
@@ -947,6 +947,8 @@ in
 
   libfx2 = with python3Packages; toPythonApplication fx2;
 
+  fitnesstrax = callPackage ../applications/misc/fitnesstrax/default.nix { pkgs = pkgs; };
```
```suggestion
  fitnesstrax = callPackage ../applications/misc/fitnesstrax/default.nix { };
```

savannidgerinel committed:
Changes from code review, + rebase onto nixos-unstable

The rebase was at the suggestion of worldofpeace in IRC. nixos-unstable
may have updates to the Rust build infrastructure that don't exist in
master. Nevertheless, this should still go into master.

Still learning all of the processes. Harder than I expected.

worldofpeace commented on pkgs/applications/misc/fitnesstrax/default.nix:
```diff
@@ -0,0 +1,51 @@
+  src = fetchFromGitHub {
+    owner = "luminescent-dreams";
+    repo = "fitnesstrax";
+    rev = "fitnesstrax-0.1.0";
```
```suggestion
    rev = "${pname}-${version}";
```

savannidgerinel committed:
Update how rev is being set in the metadata

Co-Authored-By: worldofpeace <worldofpeace@protonmail.ch>

worldofpeace committed:
fix deps

I realized after looking at http://gtk-rs.org/docs/requirements.html
that you only need gtk3 to build the gtk rust binding.
gtk3 propagates all the other required development libs.
This is why I only put gtk3 and glib.

worldofpeace commented:
I noticed something up with your deps, which I actually fixed for you https://github.com/NixOS/nixpkgs/pull/83336/commits/a9f12d5f5ab2dd733f3cf43270d87f7e121f8a68.
Only system libs you'd need would have been for `gtk-rs`.

savannidgerinel commented:
@worldofpeace I am not sure how to proceed. I had no idea that you could make a commit to a PR from a remote repository, so I don't know how to integrate this into my PR.

And, I don't understand the error that grahamcofborg-eval is reporting.

What do you advise as my next step?

worldofpeace commented:
@savannidgerinel Just pull from origin.

The error is a problem with ofborg that graham is fixing/maybe fixed.

Will try an eval
@GrahamcOfBorg build fitnesstrax

But generally, if you could build the package it should be fine.
(though the metadata could be messed up, but I don't think it is)

savannidgerinel committed:
pname doesn't actually exist

worldofpeace commented on pkgs/applications/misc/fitnesstrax/default.nix:
```diff
@@ -0,0 +1,43 @@
+  src = fetchFromGitHub {
+    owner = "luminescent-dreams";
+    repo = "fitnesstrax";
+    rev = "${name}-${version}";
```
you should switch the `name` attr to `pname` at a lower level they are concatenated like `${pname}-${version}` to make `name` But it's not available at an attribute level. Currently the derivation name is just `fitnesstrax`.

savannidgerinel replied:
done!

savannidgerinel committed:
Change name attr to pname

savannidgerinel committed:
Update Cargo checksum

worldofpeace commented:
@savannidgerinel Squash merged, thanks :sparkles:
